openhtf/core/__init__.py

```diff
@@ -114,6 +114,10 @@ def _thread_proc(self):
       # Any access to self._exit_stack must be done while holding this lock.
       with self._lock:
         self._exit_stack = exit_stack
+        # We need to exit the PhaseExecutor before we tear down the plugs
+        # that a phase may be executing, so we add them in reverse order.
+        exit_stack.callback(self.test_state.plug_manager.tear_down_plugs)
+        exit_stack.callback(executor.stop)
```

Comment:
ah, right, there should be an `exit_stack.callback(self.test_state.plug_manager.tear_down_plugs` between the `initialize_plugs()` and `executor.execute_start_trigger()`, but then it gets a bit sloppy to handle all the branching.  moving it up here is all right as long as you've double checked that both `executor.stop()` and `tear_down_plugs()` can be safely called if the test isn't started (ie, immediately upon exiting this `with self._lock:` block).